### DISCUSSION TITLE
How to decode a multipart file received from the client in ariadne wsgi resolver

### DISCUSSION THREAD

**Discussion by sumchans** [mirumee/ariadne#1093] [Q&A]
I am trying to upload a file to Azure Storage account using python and ariadne graphql. The file will be received from a flutter app in multipart form. How do I decode that multipart file and upload it to Azure Storage account using the araidne graphql schema and resolver?

I am using the wsgiMiddleWare.

Here is my schema -
```
type_defs = gql("""
    scalar Upload
    type Mutation {
        fileUpload(file: FileUpload): Result!    
    }
    input FileUpload {
        file: Upload
        fileName: String
        fileType:String
        fileExtension:String
        uploadedDate: String
        containerName: String
    }

    type Result {
        status: Int
        error: String
        user: User
        users: [User]
        fileUrl: String
    }
""")
```

And here is my resolver in python. This code currently breaks when I send in a file to the ariadne graphql server.
```
@mutation.field("fileUpload")
def resolve_fileUpload(_, info, file):
    print(file['containerName'])
    file_path = file['file'] if 'file' in file else None
    file_name = file['fileName'] if 'fileName' in file else None
    file_type = file['fileType'] if 'fileType' in file else None
    file_extension = file['fileExtension'] if 'fileExtension' in file else None
    uploaded_date = file['uploadedDate'] if 'uploadedDate' in file else None
    container_name = file['containerName'] if 'containerName' in file else None

    container_client = blob_service_client.get_container_client(
         container_name)
     if not container_client.exists():
         container_client.create_container()
         container_client.set_container_metadata(
             metadata={'Created_Date': uploaded_date})

     with open(file_path, "rb") as file:
         result = container_client.upload_blob(
             name=file_name, data=file_path)
```
Experts please advise.

**Comment by dacevedo12** [comment]
File upload with ariadne usually requires 3 things:
- The client uses https://github.com/jaydenseric/apollo-upload-client or equivalent
- Install ariadne with the `file-uploads` extra
- Declare the upload scalar in the SDL `scalar Upload` and in the `make_executable_schema` function

After that, you should be able to work with the file normally after obtaining it from the mutation `kwargs["file"]["file"]`
Details: https://ariadnegraphql.org/docs/file-uploads

**Reply by sumchans** [reply]
Hello @dacevedo12, I am using this package on the client side to send the file to the graphql server. https://pub.dev/packages/graphql_flutter

1. Do you have any idea if this package uses the apollo-upload-client? 

2 & 3 has been implemented on the ariadne side.

**Reply by dacevedo12** [reply]
I haven't worked with flutter. The important thing is this spec https://github.com/jaydenseric/graphql-multipart-request-spec

If your flutter gql client implements it, should work OK.

**Reply by sumchans** [reply]
Okay thanks @dacevedo12

**Reply by sumchans** [reply]
@dacevedo12 If we perform all the steps above could we use the file['file'] just like that in upload_blob method as in the above code? The reason why I am asking is I have read somewhere that I should be using the python-multipart package and parse the received file. Should any of that needs to be done?

**Reply by dacevedo12** [reply]
It is not a file path. the kwarg file is a FileUpload instance, which you can 'await file.read()'. 

Printing it can help you understand the type its receiving and how to work with it

**Reply by sumchans** [reply]
@dacevedo12 So I was checking on everything you mentioned and also tried out a few things. I found that the flutter client is sending in the file as an UploadFile object. 

Here is my minified python code. As you see below the print statement prints **File - <starlette.datastructures.UploadFile object at 0x000001E5C1EFCC70>**. 
```
@mutation.field("fileUpload")
async def resolve_fileUpload(_, info, file):
    print(f"File - {file}")
    bytes_file = await file.read()
    container_client = blob_service_client.get_container_client(
        '4160000000')
    if not container_client.exists():
        container_client.create_container()

    with open(file, "rb") as file:
        result = container_client.upload_blob(
            name='avatar', data=bytes_file)
    return {
        "status": 200,
        "error": "",
        "fileUrl": "www.test.com"
    }
```

Then I tried reading the contents of the file to convert it to bytes so that I can use it in the upload_blob method. This is the error I get when I try to use the converted file in the upload_blob method - **expected str, bytes or os.PathLike object, not UploadFile**

The code code runs if I comment out the with statement with the upload_blob method.

**Reply by dacevedo12** [reply]
You probably want to read the file first. Please refer to starlette docs about UploadFile for more info

**Comment by sumchans** [comment]
Okay got it. thanks again.
